compiler/sem/semexprs.nim

```diff
@@ -1872,205 +1872,220 @@ proc buildOverloadedSubscripts(n: PNode, ident: PIdent): PNode =
+      #      used instead, which will reject the call if not valid
+      n.transitionSonsKind(nkCall) # XXX: in-place modification
+      case s.kind
+      of skMacro: result = semMacroExpr(c, n, s, flags)
+      of skTemplate: result = semTemplateExpr(c, n, s, flags)
+      else: discard
+      return
+    of skType:
+      # ``typ[...]`` - explicit generic type instantiation
+      result = copyNodeWithKids(n)
+      result[0] = target
+      # XXX: error propagation is missing
+      result.typ = makeTypeDesc(c, semTypeNode(c, result, nil))
       return
-    elif arr.n != nil:
-      return semSubscript(c, arr.n, flags)
     else:
-      arr = arr.base
+      discard "something"
+
+  # set up the production
+  result = copyNodeWithKids(n)
+  result[0] = target
+
+  # perform manual overload resolution based on arity, target's type, and, if
+  # present, the first operand's type
+
+  if n.len == 1 and
+     skipTypes(target.typ,
+               {tyGenericInst, tyAlias, tySink}).kind in {tyPtr, tyRef}:
+    result.transitionSonsKind(nkDerefExpr)
+    return semDerefEval(c, result)
+
+  let arr = skipTypes(target.typ, {tyGenericInst, tyUserTypeClassInst,
+                                   tyPtr, tyRef, tyAlias, tySink})
 
   case arr.kind
   of tyArray, tyOpenArray, tyVarargs, tySequence, tyString, tyCstring,
     tyUncheckedArray:
-    if n.len != 2: return nil
-    n[0] = makeDeref(n[0])
-    for i in 1..<n.len:
-      n[i] = semExprWithType(c, n[i], flags*{efInTypeof})
-    # Arrays index type is dictated by the range's type
-    if arr.kind == tyArray:
-      var indexType = arr[0]
-      var arg = indexTypesMatch(c, indexType, n[1].typ, n[1])
-      if arg != nil:
-        n[1] = arg
-        result = n
+    if n.len == 2:
+      result[1] = semExprWithType(c, n[1], flags*{efInTypeof})
+      if arr.kind == tyArray:
+        # for arrays, the index types have to match
+        let arg = indexTypesMatch(c, arr[0], result[1].typ, result[1])
+        if arg != nil:
+          result[1] = arg
+          result.typ = elemType(arr)
+      elif result[1].typ.skipTypes(abstractRange-{tyDistinct}).kind in
+          {tyInt..tyInt64, tyUInt..tyUInt64}:
+        # other types have a bit more of leeway - all integer types are
+        # allowed for the index value
         result.typ = elemType(arr)
-    # Other types have a bit more of leeway
-    elif n[1].typ.skipTypes(abstractRange-{tyDistinct}).kind in
-        {tyInt..tyInt64, tyUInt..tyUInt64}:
-      result = n
-      result.typ = elemType(arr)
+
+      if result.typ != nil:
+        # only insert the dereferences when the built-in overload matched
+        result[0] = makeDeref(target)
+
   of tyTypeDesc:
     # The result so far is a tyTypeDesc bound
     # a tyGenericBody. The line below will substitute
     # it with the instantiated type.
-    result = n
-    result.typ = makeTypeDesc(c, semTypeNode(c, n, nil))
+    # XXX: error propagation is missing
```

Comment:
Should this `localReport` on an `nkError` in the meantime

Reply:
There's nothing to `localReport` here, unfortunately. In case of an error, `semTypeNode` already `localReport`s internally and only returns a `tyError` with no AST.

Reply:
Better than the alternative of "leaking" the error node, I suppose.